src/ChainRouterBcLayer.php

```diff
@@ -0,0 +1,124 @@
+
+$urlGeneratorParameters = (new \ReflectionClass(UrlGeneratorInterface::class))->getMethod('generate')->getParameters();
+$urlGeneratorNameParameter = array_shift($urlGeneratorParameters);
+if ($urlGeneratorNameParameter && $urlGeneratorNameParameter->hasType() && 'string' === $urlGeneratorNameParameter->getType()) {
+    /**
+     * @internal
+     */
+    class ChainRouterBcLayer extends ChainRouterBaseBcLayer
+    {
+        /**
+         * Loops through all registered routers and returns a router if one is found.
+         * It will always return the first route generated.
+         */
+        public function generate(string $name, $parameters = [], $absolute = UrlGeneratorInterface::ABSOLUTE_PATH)
+        {
+            return $this->doGenerate($name, $parameters, $absolute);
+        }
+    }
+} else {
+    /**
+     * @internal
+     */
+    class ChainRouterBcLayer extends ChainRouterBaseBcLayer
+    {
+        /**
+         * Loops through all registered routers and returns a router if one is found.
+         * It will always return the first route generated.
+         */
+        public function generate($name, $parameters = [], $absolute = UrlGeneratorInterface::ABSOLUTE_PATH)
+        {
+            if (!is_string($name)) {
+                @trigger_error(sprintf('Passing an object as the route name is deprecated in symfony-cmf/Routing v2.2 and will not work in Symfony 5.0. Pass the `RouteObjectInterface::OBJECT_BASED_ROUTE_NAME` constant as the route name and the object as "%s" parameter in the parameters array.', RouteObjectInterface::ROUTE_OBJECT), E_USER_DEPRECATED);
+
+                $parameters[RouteObjectInterface::ROUTE_OBJECT] = $name;
+                $name = RouteObjectInterface::OBJECT_BASED_ROUTE_NAME;
+            }
+
+            return $this->doGenerate($name, $parameters, $absolute);
+        }
+    }
+}
+
+/**
+ * @internal
+ */
+abstract class ChainRouterBaseBcLayer
+{
+    protected function doGenerate($name, $parameters, $absolute)
+    {
+        $debug = [];
+
+        foreach ($this->all() as $router) {
+            // if $router does not announce it is capable of handling
+            // non-string routes and $name is not a string, continue
+            if (array_key_exists(RouteObjectInterface::ROUTE_OBJECT, $parameters) && is_object($parameters[RouteObjectInterface::ROUTE_OBJECT]) && !$router instanceof VersatileGeneratorInterface) {
+                continue;
+            }
+
+            $routeName = $name;
+            if (RouteObjectInterface::OBJECT_BASED_ROUTE_NAME === $name && array_key_exists(RouteObjectInterface::ROUTE_OBJECT, $parameters) && is_object($parameters[RouteObjectInterface::ROUTE_OBJECT])) {
+                $routeName = $parameters[RouteObjectInterface::ROUTE_OBJECT];
+            }
+
+            // If $router is versatile and doesn't support this route name, continue
+            if ($router instanceof VersatileGeneratorInterface && !$router->supports($routeName)) {
```

Comment:
i think here we should pass in the route object if existing. we would need to change the phpdoc of VersatileRouterInterface, but that interface explicitly says that the argument could be an object. then routers can decide on the type of the route_object parameter whether they support generating a route for that.